test/test_functional_tensor.py

```diff
@@ -15,18 +15,31 @@
 import torchvision.transforms.functional as F
 
 
+def run_on_cpu_and_cuda(fn):
```

Comment:
This is cool, and is very similar to the PyTorch [equivalent `onlyOnCPUandCUDA`](https://github.com/pytorch/pytorch/blob/33519e19abfe99747ee654df61b83d3c82f03c27/torch/testing/_internal/common_device_type.py#L546-L556).

One thing I would like to double-check: if the test fails on CUDA (or CPU) for some reason, does the error message indicate that the issue was on CUDA? 

I think @mruberry can also give some insights / feedback on this, given that he has worked on this on PyTorch side

Reply:
This is neat but it won't indicate which (sub)test, CPU or CUDA, failed. The test itself could include the device string in error messages it prints when it fails.

This is a lightweight way of running the same test on the CPU and CUDA so maybe that's OK. A question that we may want to follow-up on is whether torchvision can just use PyTorch's device generic test framework.

Reply:
@mruberry yes, locally I quickly used [subTest](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests) but running with pytest it does not show the device neither...

Reply:
I think you could get it to work with subtest, but I think to support subtests in pytest you need a plugin? For the PyTorch CI (not sure about torchvision) we actually void using subtest for that reason.

Reply:
Actually, when test is failing on a device, test's input args are shown like that:
```
______________________________________________ Tester.test_vflip _________________________________________________________
                                                        
self = <test_functional_tensor.Tester testMethod=test_vflip>, args = (), kwargs = {}, device = 'cuda'  
```
So, I think this is OK as behaviour as we can see on which device it fails

Reply:
When a test fails, does it still run the remaining tests? I did not look into the detail but the function name looks like it executes all the tests sequentially and one failure would stop the rest from running. This can be very painful dev experience.

Also can this test run CPU and GPU tests separately? I could be wrong but, as far as I know, you cannot run GPU tests in fbcode (so far), and when you define test suites in fbcode, one needs to separate test suites on file-level.

Reply:
> When a test fails, does it still run the remaining tests? 

@mthrok it executes all `self.test_*` as usually, but it wont continue on "cuda" device case if "cpu" subtest is failed which is IMO is OK. 
Otherwise, I can split all tests into two, such that everything is explicit. 

> Also can this test run CPU and GPU tests separately? 

It adds "cuda" device is `torch.cuda.is_available()` is True, so if there is no GPUs it still can run on CPU... I agree that the name `run_on_cpu_and_cuda` can be misleading.

Let me know if you have other thoughts.

Reply:
> > When a test fails, does it still run the remaining tests?
> 
> @mthrok it executes all `self.test_*` as usually, but it wont continue on "cuda" device case if "cpu" subtest is failed which is IMO is OK.

I have an opposite opinion. Sometimes I have to iterate CI jobs to fix something. It is easier if I can see all the failure from the beginning, so that I can locate the failure points.

FYI: I have made everything separate and explicit in torchaudio for this reason. I am not advocating that torch vision should do the same approach. I do not know as much about torch vision, so I understand if there are already established customs too, which does not align with this approach. 

https://github.com/pytorch/audio/blob/master/test/torchaudio_unittest/torchscript_consistency_impl.py
https://github.com/pytorch/audio/blob/master/test/torchaudio_unittest/torchscript_consistency_cpu_test.py
https://github.com/pytorch/audio/blob/master/test/torchaudio_unittest/torchscript_consistency_cuda_test.py

> Let me know if you have other thoughts.

Even though the tests on GPU are automatically skipped if there is no CUDA available, if these tests are skipped all the time, fbcode will issue warnings. So it is ideal if CUDA tests are separated explicitly.

Reply:
@mthrok thanks for the links, torch audio unit tests look clean and well split 👍 

I agree that it would be better to split it in torchvision too at least as different tests
```
class Tester(unittest.TestCase):

    def _test_vflip(self, device):
        # implementation

    def test_vflip_cpu(self):
        self._test_vflip("cpu")

    @skipIfNoCuda
    def test_vflip_cuda(self):
        self._test_vflip("cuda")
```

> Even though the tests on GPU are automatically skipped if there is no CUDA available, if these tests are skipped all the time, fbcode will issue warnings. So it is ideal if CUDA tests are separated explicitly.

Here, I have no idea how it works, I let @fmassa to decide on that.

Reply:
About fbcode, let's not worry about that for now, and let's just skip the tests as normally